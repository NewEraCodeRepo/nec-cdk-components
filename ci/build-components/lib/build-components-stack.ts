/* tslint:disable:no-unused-expression */
import codebuild = require('@aws-cdk/aws-codebuild');
import ssm = require('@aws-cdk/aws-ssm');
import cdk = require('@aws-cdk/core');
import { IConfig } from '../../../scripts/config';

export interface IBuildComponentsStackProps extends cdk.StackProps {
  config: IConfig;
}

export class BuildComponentsStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props: IBuildComponentsStackProps) {
    super(scope, id, props);

    const {
      githubConfig: { githubOwner, githubRepo, githubBranch },
    } = props.config;

    const gitHubSource = codebuild.Source.gitHub({
      owner: githubOwner,
      repo: githubRepo,
      webhook: true, // optional, default: true if `webhookFilteres` were provided, false otherwise
      webhookFilters: [codebuild.FilterGroup.inEventOf(codebuild.EventAction.PUSH).andBranchIs(githubBranch)], // optional, by default all pushes and Pull Requests will trigger a build
    });

    const secureNpmToken = ssm.StringParameter.fromSecureStringParameterAttributes(this, 'SecureNpmToken', {
      parameterName: '/TOKENS/CODEBUILD/ReadWriteNpmToken',
      version: 1,
    });

    const componentBuildProject = new codebuild.Project(this, 'ComponentBuildProject', {
      badge: true,
      description: 'NewEraCode Component CodeBuild Project',
      environmentVariables: {
        NODE_ENV: { value: 'development' },
        NODE_OPTIONS: { value: '--max-old-space-size=4096' },
        NPM_CONFIG_LOGLEVEL: { value: 'error' },
        NPM_CONFIG_PRODUCTION: { value: 'false' },
        NPM_TOKEN: {
          type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
          value: '/TOKENS/CODEBUILD/ReadWriteNpmToken',
        },
        YARN_PRODUCTION: { value: 'false' },
      },
      projectName: 'CDKComponents',
      source: gitHubSource,
    });

    secureNpmToken.grantRead(componentBuildProject);
  }
}
