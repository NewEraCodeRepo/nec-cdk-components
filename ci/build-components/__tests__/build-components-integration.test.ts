import { App, Environment, Stack, Tag } from '@aws-cdk/core';
import 'jest-cdk-snapshot';
import { BuildComponentsStack } from '../lib/build-components-stack';

describe('cdk-deletion-protection', () => {
  test('snapshot', () => {
    const env: Environment = {
      account: '123456789012',
      region: 'us-east-1',
    };

    const config = {
      githubConfig: {
        githubBranch: 'develop',
        githubOwner: 'NewEraCode-Holdings',
        githubRepo: 'kh-cdk-components',
      },
      nodeConfig: {
        configProd: 'false',
        logLevel: 'error',
        nodeEnv: 'development',
        yarnProd: 'false',
      },
      secretParams: {
        npmToken: 'testnpmToken',
      },
      secretsManager: {
        githubToken: 'testgithubToken',
      },
    };

    // Stack description
    const description = 'Builds a CI/CD platform that publishes cdk modules to a private npm registry';

    const app = new App();
    const buildComponentsStack = new BuildComponentsStack(app, 'BuildComponentsStack', { config, env, description });
    // Add a tag to all constructs in the stack
    Tag.add(buildComponentsStack, 'Project', 'CDKComponents');

    expect(buildComponentsStack).toMatchCdkSnapshot();
  });
});
