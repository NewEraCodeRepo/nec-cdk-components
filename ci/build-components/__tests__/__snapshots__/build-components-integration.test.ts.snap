// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdk-deletion-protection snapshot 1`] = `
Object {
  "Description": "Builds a CI/CD platform that publishes cdk modules to a private npm registry",
  "Resources": Object {
    "ComponentBuildProject876E06DD": Object {
      "Properties": Object {
        "Artifacts": Object {
          "Type": "NO_ARTIFACTS",
        },
        "BadgeEnabled": true,
        "Description": "NewEraCode Component CodeBuild Project",
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "EnvironmentVariables": Array [
            Object {
              "Name": "NODE_ENV",
              "Type": "PLAINTEXT",
              "Value": "development",
            },
            Object {
              "Name": "NODE_OPTIONS",
              "Type": "PLAINTEXT",
              "Value": "--max-old-space-size=4096",
            },
            Object {
              "Name": "NPM_CONFIG_LOGLEVEL",
              "Type": "PLAINTEXT",
              "Value": "error",
            },
            Object {
              "Name": "NPM_CONFIG_PRODUCTION",
              "Type": "PLAINTEXT",
              "Value": "false",
            },
            Object {
              "Name": "NPM_TOKEN",
              "Type": "PARAMETER_STORE",
              "Value": "/TOKENS/CODEBUILD/ReadWriteNpmToken",
            },
            Object {
              "Name": "YARN_PRODUCTION",
              "Type": "PLAINTEXT",
              "Value": "false",
            },
          ],
          "Image": "aws/codebuild/standard:1.0",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "Name": "CDKComponents",
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "ComponentBuildProjectRoleECF20D25",
            "Arn",
          ],
        },
        "Source": Object {
          "Location": "https://github.com/NewEraCodeRepo/nec-cdk-components.git",
          "ReportBuildStatus": true,
          "Type": "GITHUB",
        },
        "Tags": Array [
          Object {
            "Key": "Project",
            "Value": "CDKComponents",
          },
        ],
        "Triggers": Object {
          "FilterGroups": Array [
            Array [
              Object {
                "Pattern": "PUSH",
                "Type": "EVENT",
              },
              Object {
                "Pattern": "refs/heads/develop",
                "Type": "HEAD_REF",
              },
            ],
          ],
          "Webhook": true,
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "ComponentBuildProjectRoleDefaultPolicy08D1D242": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:123456789012:log-group:/aws/codebuild/",
                      Object {
                        "Ref": "ComponentBuildProject876E06DD",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:123456789012:log-group:/aws/codebuild/",
                      Object {
                        "Ref": "ComponentBuildProject876E06DD",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "ssm:DescribeParameters",
                "ssm:GetParameters",
                "ssm:GetParameter",
                "ssm:GetParameterHistory",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:us-east-1:123456789012:parameter/TOKENS/CODEBUILD/ReadWriteNpmToken",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ComponentBuildProjectRoleDefaultPolicy08D1D242",
        "Roles": Array [
          Object {
            "Ref": "ComponentBuildProjectRoleECF20D25",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ComponentBuildProjectRoleECF20D25": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "Project",
            "Value": "CDKComponents",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
